File: server/handlers/constructions.go
package handlers

import (
	"encoding/json"
	"net/http"
	"server/db"
	"strings"
	"time"
)

// ── Models ──────────────────────────────────────────────────────────────────

type Construction struct {
	ID                   int             `json:"id"`
	DesignMode           string          `json:"design_mode"`
	FurnitureType        string          `json:"furniture_type"`
	PrimaryFinish        string          `json:"primary_finish"`
	SecondaryFinish      string          `json:"secondary_finish"`
	ReferenceCode        string          `json:"reference_code"`
	CustomerRequirements string          `json:"customer_requirements"`
	ExtraFinish          bool            `json:"extra_finish"`
	SpecialFinishes      json.RawMessage `json:"special_finishes"`
	ImageURL             string          `json:"image_url"`
	RequestDate          time.Time       `json:"request_date"`
}

type CreateConstructionRequest struct {
	DesignMode           string   `json:"design_mode"`
	FurnitureType        string   `json:"furniture_type"`
	PrimaryFinish        string   `json:"primary_finish"`
	SecondaryFinish      string   `json:"secondary_finish"`
	ReferenceCode        string   `json:"reference_code"`
	CustomerRequirements string   `json:"customer_requirements"`
	ExtraFinish          bool     `json:"extra_finish"`
	SpecialFinishes      []string `json:"special_finishes"`
	ImageURL             string   `json:"image_url"`
}

// ── Handler ──────────────────────────────────────────────────────────────────

func ConstructionsHandler(w http.ResponseWriter, r *http.Request) {
	switch r.Method {
	case http.MethodGet:
		getConstructions(w, r)
	case http.MethodPost:
		createConstruction(w, r)
	default:
		http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
	}
}

func getConstructions(w http.ResponseWriter, r *http.Request) {
	rows, err := db.DB.Query(`
		SELECT id, COALESCE(design_mode, 'OEM'), furniture_type, primary_finish, COALESCE(secondary_finish,''),
		       COALESCE(reference_code,''), COALESCE(customer_requirements,''),
		       extra_finish, COALESCE(special_finishes, '[]'), COALESCE(image_url,''), request_date
		FROM constructions ORDER BY request_date DESC
	`)
	if err != nil {
		jsonError(w, err.Error(), http.StatusInternalServerError)
		return
	}
	defer rows.Close()

	var constructions []Construction
	for rows.Next() {
		var c Construction
		var extraFinish int
		var specialFinishesStr string
		if err := rows.Scan(&c.ID, &c.DesignMode, &c.FurnitureType, &c.PrimaryFinish, &c.SecondaryFinish,
			&c.ReferenceCode, &c.CustomerRequirements,
			&extraFinish, &specialFinishesStr, &c.ImageURL, &c.RequestDate); err != nil {
			continue
		}
		c.ExtraFinish = extraFinish == 1
		c.SpecialFinishes = json.RawMessage(specialFinishesStr)
		constructions = append(constructions, c)
	}
	if constructions == nil {
		constructions = []Construction{}
	}
	writeJSON(w, http.StatusOK, constructions)
}

func createConstruction(w http.ResponseWriter, r *http.Request) {
	var req CreateConstructionRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		jsonError(w, "Invalid request body", http.StatusBadRequest)
		return
	}
	if req.FurnitureType == "" || req.PrimaryFinish == "" {
		jsonError(w, "furniture_type and primary_finish are required", http.StatusBadRequest)
		return
	}
	req.DesignMode = strings.TrimSpace(req.DesignMode)
	req.FurnitureType = strings.TrimSpace(req.FurnitureType)
	req.PrimaryFinish = strings.TrimSpace(req.PrimaryFinish)
	req.SecondaryFinish = strings.TrimSpace(req.SecondaryFinish)
	req.ReferenceCode = strings.TrimSpace(req.ReferenceCode)
	req.CustomerRequirements = strings.TrimSpace(req.CustomerRequirements)
	if req.DesignMode == "" {
		req.DesignMode = "OEM"
	}
	validModes := map[string]bool{"OEM": true, "ODM": true, "Bespoke": true}
	if !validModes[req.DesignMode] {
		jsonError(w, "design_mode must be OEM, ODM, or Bespoke", http.StatusBadRequest)
		return
	}

	specialFinishesJSON, err := json.Marshal(req.SpecialFinishes)
	if err != nil {
		jsonError(w, "Invalid special_finishes", http.StatusBadRequest)
		return
	}

	result, err := db.DB.Exec(
		`INSERT INTO constructions (design_mode, furniture_type, primary_finish, secondary_finish, reference_code, customer_requirements, extra_finish, special_finishes, image_url)
		 VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)`,
		req.DesignMode, req.FurnitureType, req.PrimaryFinish, req.SecondaryFinish, req.ReferenceCode, req.CustomerRequirements,
		req.ExtraFinish, string(specialFinishesJSON), req.ImageURL,
	)
	if err != nil {
		jsonError(w, err.Error(), http.StatusInternalServerError)
		return
	}
	id, _ := result.LastInsertId()
	writeJSON(w, http.StatusCreated, map[string]interface{}{"message": "Design specification created", "id": id})
}

File: serverJS/handlers/constructions.js
import { pool } from '../db/connection.js';
import { writeJSON, jsonError } from '../utils/response.js';

export async function constructionsHandler(req, res) {
  try {
    if (req.method === 'GET') {
      return getConstructions(req, res);
    } else if (req.method === 'POST') {
      return createConstruction(req, res);
    } else {
      return res.status(405).json({ error: 'Method not allowed' });
    }
  } catch (error) {
    console.error('Constructions handler error:', error);
    return jsonError(res, error.message, 500);
  }
}

async function getConstructions(req, res) {
  try {
    const connection = await pool.getConnection();
    try {
      const [rows] = await connection.execute(`
        SELECT id, COALESCE(design_mode, 'OEM') AS design_mode, furniture_type, primary_finish, COALESCE(secondary_finish,'') AS secondary_finish,
               COALESCE(reference_code,'') AS reference_code, COALESCE(customer_requirements,'') AS customer_requirements,
               extra_finish, COALESCE(special_finishes, '[]') AS special_finishes, COALESCE(image_url,'') AS image_url, request_date
        FROM constructions ORDER BY request_date DESC
      `);

      const constructions = rows.map(row => ({
        id: row.id,
        design_mode: row.design_mode,
        furniture_type: row.furniture_type,
        primary_finish: row.primary_finish,
        secondary_finish: row.secondary_finish,
        reference_code: row.reference_code,
        customer_requirements: row.customer_requirements,
        extra_finish: Number(row.extra_finish) === 1,
        special_finishes: (() => {
          try {
            const parsed = typeof row.special_finishes === 'string'
              ? JSON.parse(row.special_finishes || '[]')
              : row.special_finishes;
            return Array.isArray(parsed) ? parsed : [];
          } catch {
            return [];
          }
        })(),
        image_url: row.image_url,
        request_date: row.request_date
      }));

      return writeJSON(res, 200, constructions);
    } finally {
      connection.release();
    }
  } catch (error) {
    console.error('Get constructions error:', error);
    return jsonError(res, error.message, 500);
  }
}

async function createConstruction(req, res) {
  try {
    const { design_mode, furniture_type, primary_finish, secondary_finish, reference_code, customer_requirements, extra_finish, special_finishes, image_url } = req.body;

    if (!furniture_type || !primary_finish) {
      return jsonError(res, 'furniture_type and primary_finish are required', 400);
    }

    let finalDesignMode = (design_mode || '').trim() || 'OEM';
    const validModes = ['OEM', 'ODM', 'Bespoke'];
    if (!validModes.includes(finalDesignMode)) {
      return jsonError(res, 'design_mode must be OEM, ODM, or Bespoke', 400);
    }

    const specialFinishesJSON = JSON.stringify(special_finishes || []);

    const connection = await pool.getConnection();
    try {
      const [result] = await connection.execute(`
        INSERT INTO constructions 
        (design_mode, furniture_type, primary_finish, secondary_finish, reference_code, customer_requirements, extra_finish, special_finishes, image_url)
        VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)
      `, [
        finalDesignMode,
        (furniture_type || '').trim(),
        (primary_finish || '').trim(),
        (secondary_finish || '').trim(),
        (reference_code || '').trim(),
        (customer_requirements || '').trim(),
        extra_finish ? 1 : 0,
        specialFinishesJSON,
        (image_url || '').trim()
      ]);

      return writeJSON(res, 201, {
        message: 'Design specification created',
        id: result.insertId
      });
    } finally {
      connection.release();
    }
  } catch (error) {
    console.error('Create construction error:', error);
    return jsonError(res, error.message, 500);
  }
}
